[Feature] Export 3D scene feature

NEDJIMAbelgacem opened the pull request:
## Description

This PR is still under development and a part of my summer of code project (sorry if it's too big for reviews :( ).
I started working on exporting 3D map scenes to ,obj files.
Currently in my implementation, the user writes scene name (which will be the .obj file name later) and chooses a folder to export the 3D scene files to, I did it like this because later I will add textures and .mtl files that will contain material information.
The user can also give an integer to specify the resolution of the terrain and specify whether the triangles should look smooth in an app like blender.
Here is my initial UI for the export settings:
 ![map_export_ui](https://user-images.githubusercontent.com/29183781/86510289-58e8c900-bde6-11ea-89e1-2e1b7d5dc057.png)
I made a separate class to manage the export settings because we will have more settings later on.
Here is a screenshot of a terrain and how it is exported to blender with the smoothing option on and off:
Original 3D view:
![image](https://user-images.githubusercontent.com/29183781/86510370-fc39de00-bde6-11ea-9247-f3c1d183c157.png)
smoothing option off:
![image](https://user-images.githubusercontent.com/29183781/86510424-66528300-bde7-11ea-9f89-37de70dce86e.png)
smoothing option on:
![image](https://user-images.githubusercontent.com/29183781/86510446-82eebb00-bde7-11ea-9188-6bdef17a3edf.png)
The exporter can also handle exporting vector layers and flat terrains:
![vector_layer_and_flat_terrain](https://user-images.githubusercontent.com/29183781/86510462-ae71a580-bde7-11ea-8738-c2fd1debd2ef.png)

I tried to make well written code focusing on extensibility.
I want to know if the architecture of classes is good enough, if anyone have suggestion to improve the quality please let me know. Any feedback or suggestion are welcome.

TODO:
- [x] Find good icon for scene exporting action.
- [x] Add normals
- [x] Add texture coordinates
- [x] Manage copying the textures from the scene to a file.
- [x] Handle .mtl specification.
- [x] Add phong material settings handling.
- [ ] Exporting of other types of layers.
- [ ] Permit the user to choose which layers he wants to export and configure them individually.

NEDJIMAbelgacem committed:
initial implementation of the scene exporter class

NEDJIMAbelgacem committed:
Got exporting feature to work on vector layer data

NEDJIMAbelgacem committed:
Terrain export initial implementation

NEDJIMAbelgacem committed:
Generate the flat terrain all at once

NEDJIMAbelgacem committed:
visble DEM terrain export

NEDJIMAbelgacem committed:
Not loaded DEM terrain chunks export

NEDJIMAbelgacem committed:
Fixed invalid vertex Y coordinate problem

NEDJIMAbelgacem committed:
Added initial scene export dialog and smooth edges parameter

NEDJIMAbelgacem committed:
Got the scene to export using a selected level of details

NEDJIMAbelgacem committed:
Created export settings class

NEDJIMAbelgacem committed:
Created QgsExportObject to manage each object data (position coordinates, textures, normals ...)
Fixed an issue with QgsDemHeightMapGenerator::renderSynchronously because it wasm't doing the same things as asynchronous rendering function
Switched to using just one tile for DEM terrain and using a resolution parameter to handle level of details

NEDJIMAbelgacem committed:
Adding documentation and fixing code layout

NEDJIMAbelgacem committed:
Fixing documentation

NEDJIMAbelgacem committed:
Fixed documentation

nyalldawson commented on src/3d/qgs3dmapexportsettings.cpp:
```diff
@@ -0,0 +1,26 @@
+Qgs3DMapExportSettings::Qgs3DMapExportSettings( QObject *parent )
+  : QObject( parent )
+  , mSceneName( "Scene" )
+  , mSceneFolderPath( "~/" )
```
You can use QDir::homePath() for this, and it will work nicely on windows

nyalldawson commented on src/3d/qgs3dmapexportsettings.cpp:
```diff
@@ -0,0 +1,26 @@
+  : QObject( parent )
+  , mSceneName( "Scene" )
+  , mSceneFolderPath( "~/" )
+  , mTerrainResolution( 128 )
```
These initializers should be moved to the header (C++11) style

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -763,3 +766,23 @@ void Qgs3DMapScene::updateSceneState()
+
+  // TODO: make the scene .obj file created using a function in Qgs3DMapExportSettings
+  QString filePath = exportSettings.sceneFolderPath();
+  if ( !filePath.endsWith( '/' ) ) filePath.push_back( '/' );
```
Use the QDir API to construct the file path

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,493 @@
+void Qgs3DSceneExporter::saveToFile( const QString &filePath )
+{
+  QFile file( filePath );
+  if ( !file.open( QIODevice::WriteOnly | QIODevice::Text ) )
```
Add the truncate flag here too, so that fine overwrites work correctly

nyalldawson commented on src/3d/qgsexportobject.h:
```diff
@@ -0,0 +1,71 @@
+ * \ingroup 3d
+ * \since QGIS 3.16
+ */
+class QgsExportObject : public QObject
```
This class name is quite generic -- would Qgs3DExportObject work?

nyalldawson commented on src/ui/3d/map3dexportwidget.ui:
```diff
@@ -0,0 +1,89 @@
+      </widget>
+     </item>
+     <item row="0" column="1" colspan="2">
+      <widget class="QLineEdit" name="sceneNameLineEdit"/>
```
You should use QgsFileWidget here, that's got lots of built in logic for selection of folders

nyalldawson reviewed (commented):
Nice work! Some minor comments on the general coding here (not a review of the 3d component, I'll let Martin do that!), but nothing major

NEDJIMAbelgacem committed:
Improved the code according to Nyall's suggessions

NEDJIMAbelgacem committed:
Fixed flat terrain scale bug

NEDJIMAbelgacem commented:
> Nice work! Some minor comments on the general coding here (not a review of the 3d component, I'll let Martin do that!), but nothing major

Thanks!
feel free to give me your thoughts of the code structure even if that's not the a full review it will help me fix things before the code gets even bigger

NEDJIMAbelgacem committed:
Added normals export

NEDJIMAbelgacem committed:
Merge branch 'master' into export-scene-feature

NEDJIMAbelgacem committed:
texture coordinates generation
terrain texture generation
for flat and DEM terrains

NEDJIMAbelgacem committed:
Added missing documentation

NEDJIMAbelgacem committed:
Added terrain texture resolution spinbox

NEDJIMAbelgacem committed:
Added initial mtl implementation

NEDJIMAbelgacem committed:
Added scale parameter

wonder-sk commented on src/3d/CMakeLists.txt:
```diff
@@ -110,6 +111,9 @@ SET(QGIS_3D_HDRS
   chunks/qgschunkqueuejob_p.h
   mesh/qgsmeshterraingenerator.h
+  qgs3dsceneexporter.h
+  qgs3dmapexportsettings.cpp
```
cpp file within headers section?

NEDJIMAbelgacem replied:
Sorry I added it by mistake and forgot it there :facepalm:

wonder-sk commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,159 @@
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, const QVector<unsigned int> &faceIndex, float scale, const QVector3D translation )
+{
+  // TODO: delete vertices that are not used
```
Does it happen that we have vertices that do not get used?

NEDJIMAbelgacem replied:
I don't think that happens actually, I wrote this when I was looking for why invalid vertices are returned. I think that could happen if the height map is not square and we end up storing vertex coordinates for invalid faces even though we don't actually use it.
This code here detects invalid faces:
```
    if ( faceIndex[i] == faceIndex[i + 1] && faceIndex[i + 1] == faceIndex[i + 2] ) continue;
```

wonder-sk commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,159 @@
+
+  // smoothen edges
+  if ( mSmoothEdges ) out << "s on\n";
+  else out << "s off\n";
```
code style: it's strongly preferred in qgis code to have if/else statements on separate lines:
```
if ( a )
  do_a();
else
  do_b();
```

wonder-sk commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,159 @@
+  {
+    for ( int j = 0; j < 3; ++j )
+    {
+      mVertxPosition << positionsBuffer[i + j] * scale + translation[j];
```
looks like a typo mVertxPosition -> mVertexPosition

wonder-sk commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,159 @@
+
+  auto getVertexIndex = [&]( int i ) -> QString
+  {
+    int negativeIndex = -1 - ( verticesCount - i );
```
why are the indices negative here? something that .obj format requires?

NEDJIMAbelgacem replied:
In the .obj spec negative indices refer to the last defined vertices (-1 for the last vertice).
This helps keeping each object data separated

wonder-sk commented on src/3d/qgs3dmapexportsettings.h.bom:
```diff
@@ -0,0 +1,45 @@
+﻿/***************************************************************************
```
looks like this file has been added by mistake?

NEDJIMAbelgacem replied:
Yes I deleted the file

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+}
+
+template<>
+QVector<unsigned int> getAttributeData<unsigned int>( Qt3DRender::QAttribute *attribute )
```
Maybe it would be better to rename this function? From what I understand it has different semantics that the original template function getAttributeData() which also iterates over vertices, while this one does not ... assuming it is used for index buffer?

NEDJIMAbelgacem replied:
Yes I renamed it to getIndexData

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+  return result;
+}
+
+QVector<float> createPlaneVertexData( float w, float h, const QSize &resolution )
```
do we need these createPlane*Data() methods? Can't we just extract the contents of attributes of the plane geometry created for terrain tiles?

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+  // Just create a new tile (we don't need to export exact level of details as in the scene)
+  // create the entity synchronously and then it will be deleted once our scene exporter instance is deallocated
+  const Qgs3DMapSettings &map = terrain->map3D();
+  QgsDemTerrainGenerator *generator = static_cast<QgsDemTerrainGenerator *>( map.terrainGenerator() );
```
I think it would be safer to make your own copy of the terrain generator... imagine that during the export there may be loading of terrain tiles in a background thread - and we may modify the resolution of that generator while it is doing some work.

NEDJIMAbelgacem replied:
I didn't think about that. I added ->clone()

NEDJIMAbelgacem replied:
At first I carelessly added ->clone() but that actually doesn't work because the height map generator gets its terrain resolution parameter from map.terrainGenerator()
This is the code from QgsDemTerrainTileLoader constructor:
```
QgsDemTerrainGenerator *generator = static_cast<QgsDemTerrainGenerator *>( map.terrainGenerator() );
heightMapGenerator = generator->heightMapGenerator();
mSkirtHeight = generator->skirtHeight();
```

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+  QgsTerrainTextureGenerator *textureGenerator = terrain->textureGenerator();
+  QSize oldResolution = textureGenerator->textureSize();
+  textureGenerator->setTextureSize( QSize( mTerrainTextureResolution, mTerrainTextureResolution ) );
+  switch ( generator->type() )
```
I think it may be a cleaner solution to add a new virtual function to QgsTerrainGenerator class, which run the export code specific to each generator. That way, all the details would stay together with the other code of that generator...

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+    {
+      transform = transformTyped;
+    }
+  }
```
It would be good to have some utility functions e.g. findGeometryRendererComponent(), findTransformComponent() to make code shorter here and in parseDemTile()...

NEDJIMAbelgacem replied:
I made them into one function template named findTypedComponent

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp.bom:
```diff
@@ -0,0 +1,522 @@
+﻿/***************************************************************************
```
This file was probably checked in by mistake

wonder-sk commented on src/3d/qgstessellatedpolygongeometry.h:
```diff
@@ -79,12 +81,15 @@ class QgsTessellatedPolygonGeometry : public Qt3DRender::QGeometry
     Qt3DRender::QAttribute *mNormalAttribute = nullptr;
     Qt3DRender::QBuffer *mVertexBuffer = nullptr;
 
+    QVector<float> mVertexData;
```
do we need this new member?

wonder-sk commented on src/3d/terrain/qgsdemterraintileloader_p.cpp:
```diff
@@ -52,7 +52,7 @@ static void _heightMapMinMax( const QByteArray &heightMap, float &zMin, float &z
 
 
-QgsDemTerrainTileLoader::QgsDemTerrainTileLoader( QgsTerrainEntity *terrain, QgsChunkNode *node )
+QgsDemTerrainTileLoader::QgsDemTerrainTileLoader( QgsTerrainEntity *terrain, QgsChunkNode *node, bool loadSynchronously )
```
Hmm I would like to avoid having a new `loadSynchornously` flag in terrain tile loaders. I would suggest that instead we have a virtual method `waitForFinished()` that would block until the loading has not finished. This should keep the code cleaner and easier to test because there will be fewer possible code paths.

wonder-sk commented on src/3d/terrain/qgsterraintextureimage_p.h:
```diff
@@ -62,6 +62,8 @@ class QgsTerrainTextureImage : public Qt3DRender::QAbstractTextureImage
     //! Returns debug information (normally map tile coordinates)
     QString imageDebugText() const { return mDebugText; }
 
+    QImage *getImage() { return &mImage; }
```
is this used anywhere?

wonder-sk commented on src/3d/terrain/qgsterraintileloader_p.cpp:
```diff
@@ -64,6 +64,11 @@ void QgsTerrainTileLoader::loadTexture()
   mTextureJobId = mTerrain->textureGenerator()->render( mExtentMapCrs, mNode->tileId(), mTileDebugText );
 }
 
+void QgsTerrainTileLoader::loadTextureSynchronously()
```
As suggested above - I would like to avoid this synchronous mode altogether and have waitForFinished() instead which would handle that.

wonder-sk commented on src/app/3d/qgs3dmapcanvasdockwidget.cpp:
```diff
@@ -291,6 +297,34 @@ void Qgs3DMapCanvasDockWidget::configure()
+  connect( &dlg, &QDialog::accepted, [ = ]()
+  {
+    w->exportScene();
+  } );
```
It would be easier to read if instead of having this slot with a lambda function the `w->exportScene()` call would be on the last line of this function... because if dlg.exec() returns true, that means the dialog has been accepted.

NEDJIMAbelgacem replied:
For some reason I thought that we will want the dialog to not block the whole app in the future so I made it like that

wonder-sk commented on src/app/3d/qgs3dmapcanvasdockwidget.cpp:
```diff
@@ -291,6 +297,34 @@ void Qgs3DMapCanvasDockWidget::configure()
+  dlg.setMinimumSize( 380, 460 );
+//  QgsGui::instance()->enableAutoGeometryRestore( &dlg );
+
+  Qgs3DMapExportSettings *exportSettings = new Qgs3DMapExportSettings( &dlg );
```
Is this deleted anywhere? Maybe better to just use stack-based local variable: `Qgs3DMapExportSettings exportSettings( &dlg );`

NEDJIMAbelgacem replied:
I made the dialog object a parent of that class, so it is deleted when the dlg object is deallocated. I guess I was over engineering here.

wonder-sk commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,524 @@
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getDemTerrainEntity( terrain, node );
+        this->parseDemTile( terrainTile, textureGenerator );
```
Are these tiles getting deleted anywhere?

NEDJIMAbelgacem replied:
Yes I made the Qgs3DSceneExporter class derive from a QObject so I can make every generated object a child of the scene exporter and it will be deleted automatically after.
It's like I made it into a policy in my mind whenever I make a new object, it should be a child of the object where it is created. That seems like a good idea for our use case.

wonder-sk reviewed (commented):
Nice work!

NEDJIMAbelgacem committed:
Some code cleanup

NEDJIMAbelgacem committed:
CMakeLists cleanup
Deleted .bom files
Qgs3DMapCanvasDockWidget::exportScene cleanup
Deleted unused functions and variables
Fixed typo

NEDJIMAbelgacem committed:
Deleted createPlane*Data functions
Switched to buffer data generator

NEDJIMAbelgacem committed:
Deleted the synchronous loading of terrain and texture
Now uses waitForFinished function

NEDJIMAbelgacem committed:
clone terrain generator before editing the resolution
set terrain in clone function
disconnect future watcher connections properly

NEDJIMAbelgacem committed:
Deleted LoD leafs from scene exporter

NEDJIMAbelgacem committed:
Set 3d.svg as export icon

NEDJIMAbelgacem committed:
Save and reload export settings

NEDJIMAbelgacem committed:
Warning message for not parsed layers

NEDJIMAbelgacem committed:
Changed vector layer parsing function

NEDJIMAbelgacem committed:
Added buffered lines export

NEDJIMAbelgacem committed:
Export lines phong material

NEDJIMAbelgacem committed:
Export polygons phong material

NEDJIMAbelgacem committed:
Select directories in the file widget

NEDJIMAbelgacem committed:
export instanced regular geometries

NEDJIMAbelgacem committed:
export the material of instanced regular geometries

NEDJIMAbelgacem committed:
Export normals of instanced regular geometries

NEDJIMAbelgacem committed:
Add smooth edges parameter where it is missing

NEDJIMAbelgacem committed:
Scene loader geometries extraction

NEDJIMAbelgacem committed:
Export meshes loaded using QMesh class

NEDJIMAbelgacem committed:
Refactor phong material setup

NEDJIMAbelgacem committed:
Refactoring

elpaso commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -92,6 +80,18 @@ void Qgs3DExportObject::setupTextureCoordinates( const QVector<float> &texturesB
+  QColor specular = material.specular();
+  QColor ambient = material.ambient();
+  float shininess = material.shininess();
+  setMaterialParameter( QString( "Kd" ), QString( "%1 %2 %3" ).arg( diffuse.redF() ).arg( diffuse.greenF() ).arg( diffuse.blueF() ) );
```
```suggestion
  setMaterialParameter( QStringLiteral( "Kd" ), QStringLiteral( "%1 %2 %3" ).arg( diffuse.redF() ).arg( diffuse.greenF() ).arg( diffuse.blueF() ) );
```

For const rvalue strings it's better and faster to use `QStringLiteral` instead of `QString`, can you please check other occurrences too?

NEDJIMAbelgacem replied:
Okay, thanks for the suggestion. I will try changing that whenever I notice myself doing it.
By the way should QStringLiteral be used for QString and const char* comparison too?

NEDJIMAbelgacem committed:
rule based renderer initial implementation

NEDJIMAbelgacem committed:
Code refactoring

NEDJIMAbelgacem committed:
Export mesh terrain position data

NEDJIMAbelgacem committed:
trying to export lines and points
refactoring

NEDJIMAbelgacem committed:
Merge branch 'master' into export-scene-feature

NEDJIMAbelgacem committed:
vecctor layer texture export

NEDJIMAbelgacem committed:
Fixed highlight lines export bug

NEDJIMAbelgacem committed:
qDebug & QString cleanup

NEDJIMAbelgacem committed:
Added QgsImageTexture documentations

NEDJIMAbelgacem committed:
Added QgsImageTexture docs

NEDJIMAbelgacem committed:
Add layer name to .obj file object name

NEDJIMAbelgacem committed:
Merge master - adapt to QgsAbstractMaterialSettings implementation

NEDJIMAbelgacem committed:
add SIP_FACTO to QgsAbstractMaterialSettings::toExportParameters

NEDJIMAbelgacem committed:
Fix documentation

nyalldawson commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,205 @@
+  : QObject( parent )
+  , mName( name )
+  , mParentName( parentName )
+  , mSmoothEdges( false )
```
Move initialization to header (c++11 style)

nyalldawson commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,205 @@
+{
+}
+
+void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, float scale, const QVector3D translation )
```
```suggestion
void Qgs3DExportObject::setupPositionCoordinates( const QVector<float> &positionsBuffer, float scale, const QVector3D& translation )
```

nyalldawson commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,205 @@
+
+void Qgs3DExportObject::setupMaterial( QgsAbstractMaterialSettings *material )
+{
+  QMap<QString, QString> parameters = material->toExportParameters();
+  for ( QString &key : parameters.keys() ) setMaterialParameter( key, parameters[key] );
```
```suggestion
  const QMap<QString, QString> parameters = material->toExportParameters();
  for ( auto it = parameters.begin(); it != parameters.end(); ++it )
  {
     setMaterialParameter( it.key(), it.value() );
  }
```

(this is considerable more efficient -- it avoids the initial loop over the container to create the keys list, and then the n x map lookups for each key in the map)

nyalldawson commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,205 @@
+
+  if ( !hasNormals && !mNormals.empty() )
+  {
+    qDebug() << "WARNING: vertex normals count and vertex positions count are different";
```
Use QgsDebugMsg instead of qDebug()

nyalldawson commented on src/3d/qgs3dexportobject.cpp:
```diff
@@ -0,0 +1,205 @@
+  for ( QString key : mMaterialParameters.keys() )
+  {
+    mtlOut << "\t" << key << " " << mMaterialParameters[key] << "\n";
+    qDebug() << "\t" << key << " " << mMaterialParameters[key] << "\n";
```
```suggestion
    QgsDebugMsgLevel( QStringLiteral("\t%1 %2").arg( key, mMaterialParameters[key] ), 3)";
```

nyalldawson commented on src/3d/qgs3dexportobject.h:
```diff
@@ -0,0 +1,123 @@
+ */
+class Qgs3DExportObject : public QObject
+{
+    Q_OBJECT
```
Why is this a QObject?

nyalldawson commented on src/3d/qgs3dmapexportsettings.h:
```diff
@@ -0,0 +1,86 @@
+ */
+class _3D_EXPORT Qgs3DMapExportSettings : public QObject
+{
+    Q_OBJECT
```
This shouldn't be a QObject either (unless I'm missing something?)

NEDJIMAbelgacem replied:
I made the Qgs3DMapExportSettings behave as a parent of  Qgs3DExportObject so that every Qgs3DExportObject instantiated using a Qgs3DMapExportSettings will be deallocated once the parent Qgs3DMapExportSettings is deleted.

nyalldawson replied:
That's using a sledgehammer approach :laugh: If you're not using signals/slots/some complex shared ownership, it's much clearer to just use plain old pointers and clean them up in the destructor (or, the modern approach of a std::vector containing std::unique_ptr, so that they'll be cleaned up automatically)

NEDJIMAbelgacem replied:
Yes this class shouldn't be QObject. It is only necessary for the Qgs3DSceneExporter to inherit from Qt3DCore::QEntity so that every QEntity we create dynamically will be deallocated properly

nyalldawson commented on src/3d/qgs3dmapexportsettings.h:
```diff
@@ -0,0 +1,86 @@
+
+/**
+ * \brief The Qgs3DMapExportSettings class
+ * Manages the various settings the user can choose from when exorting a 3D scene
```
```suggestion
 * Manages the various settings the user can choose from when exporting a 3D scene
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+  exporter.setTerrainTextureResolution( exportSettings.terrainTextureResolution() );
+  exporter.setScale( exportSettings.scale() );
+
+  for ( QgsMapLayer *layer : mLayerEntities.keys() )
+  {
+    QgsMapLayerType layerType =  layer->type();
+    Qt3DCore::QEntity *rootEntity = mLayerEntities[layer];
```
```suggestion
  for ( auto it = mLayerEntities.constBegin(); it != mLayerEntities.constEnd(); ++it )
  {
    QgsMapLayer* layer = it.key();
    QgsMapLayerType layerType =  layer->type();
    Qt3DCore::QEntity *rootEntity = it.value();
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+      case QgsMapLayerType::PluginLayer:
+        notParsedLayers.push_back( layer->name() );
+        break;
+      case QgsMapLayerType::MeshLayer:      //!< Added in 3.2
```
```suggestion
      case QgsMapLayerType::MeshLayer:
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+      case QgsMapLayerType::MeshLayer:      //!< Added in 3.2
+        notParsedLayers.push_back( layer->name() );
+        break;
+      case QgsMapLayerType::VectorTileLayer: //!< Added in 3.14
```
```suggestion
      case QgsMapLayerType::VectorTileLayer:
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+    }
+  }
+
+  if ( mTerrain != nullptr )
```
```suggestion
  if ( mTerrain )
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+
+  exporter.save( exportSettings.sceneName(), exportSettings.sceneFolderPath() );
+
+  if ( notParsedLayers.size() != 0 )
```
```suggestion
  if ( !notParsedLayers.empty() )
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+
+  if ( notParsedLayers.size() != 0 )
+  {
+    QString message = "The following layers were not exported: \n";
```
```suggestion
    QString message = QObject::tr("The following layers were not exported: ") + '\n';
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+  if ( notParsedLayers.size() != 0 )
+  {
+    QString message = "The following layers were not exported: \n";
+    for ( QString layerName : notParsedLayers )
```
```suggestion
    for ( const QString &layerName : notParsedLayers )
```

nyalldawson commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -856,3 +860,54 @@ void Qgs3DMapScene::updateSceneState()
+    QString message = "The following layers were not exported: \n";
+    for ( QString layerName : notParsedLayers )
+      message += layerName + "\n";
+    QgsMessageOutput::showMessage( "3D exporter warning", message, QgsMessageOutput::MessageText );
```
```suggestion
    QgsMessageOutput::showMessage( tr("3D exporter warning"), message, QgsMessageOutput::MessageText );
```

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,804 @@
+
+  if ( bytesStride == 0 )
+  {
+    qDebug() << "WARNING: bytesStride==0 at " << __FILE__ << ":" << __LINE__;
```
qDebug() -> QgsDebugMsg

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,804 @@
+  {
+    for ( unsigned int j = 0; j < vertexSize * sizeof( T ); j += sizeof( T ) )
+    {
+      // maybe a problem with indienness can happen?
```
```suggestion
      // maybe a problem with endianess can happen?
```

NEDJIMAbelgacem replied:
I forgot I wrote this and was like "why did I write this" :laughing:

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,804 @@
+    case Qt3DRender::QAttribute::VertexBaseType::UnsignedByte:
+      return _getIndexDataImplementation<uchar>( data );
+    default:
+      qDebug() << "WARNING: Probably trying to get index data using an attribute that has vertex data at " << __FILE__ << ":" << __LINE__;
```
QgsDebugMsg (elsewhere here too)

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,804 @@
+
+Qgs3DSceneExporter::Qgs3DSceneExporter( Qt3DCore::QNode *parent )
+  : Qt3DCore::QEntity( parent )
+  , mSmoothEdges( false )
+  , mTerrainResolution( 128 )
+  , mExportNormals( true )
+  , mExportTextures( false )
+  , mTerrainTextureResolution( 512 )
+  , mScale( 1.0f )
```
Move initialization to header

nyalldawson commented on src/3d/qgs3dsceneexporter.cpp:
```diff
@@ -0,0 +1,804 @@
+      QgsVectorLayer3DRenderer *vectorLayerRenderer = dynamic_cast< QgsVectorLayer3DRenderer *>( abstractVectorRenderer );
+      const QgsAbstract3DSymbol *symbol = vectorLayerRenderer->symbol();
+      QString symbolType = symbol->type();
+      if ( symbolType == "polygon" )
```
This logic should probably be moved to the symbol subclasses themselves

nyalldawson commented on src/3d/qgs3dsceneexporter.h:
```diff
@@ -0,0 +1,138 @@
+    QString getObjectName( const QString &name );
+  private:
+    QMap<QString, int> usedObjectNamesCounter;
+    QVector<Qgs3DExportObject *> mObjects;
```
I can't see where mObjects is cleaned up -- I believe the contents of this vector are leaked. Perhaps a destructor should be added which deletes all these objects?

NEDJIMAbelgacem replied:
I made each object inside mObjects a child of QgsSceneExporter (that's why I made the class inherit QObject).
Now I changed that so that mObjects vector are deleted inside the destructor

nyalldawson commented on src/3d/qgsabstractmaterialsettings.h:
```diff
@@ -115,6 +115,11 @@ class _3D_EXPORT QgsAbstractMaterialSettings SIP_ABSTRACT
+    /**
+     * Returns the parameters to be exported to .mtl file
+     */
+    virtual QMap<QString, QString> toExportParameters() const = 0 SIP_FACTORY;
```
```suggestion
    virtual QMap<QString, QString> toExportParameters() const = 0;
```

The factory annotation is only required for methods which return a pointer along with ownership (or a list of pointers). This function is returning values directly, so it's not required here.

nyalldawson commented on src/3d/qgsphongmaterialsettings.cpp:
```diff
@@ -141,6 +143,16 @@ QgsLineMaterial *QgsPhongMaterialSettings::toLineMaterial( const QgsMaterialCont
+QMap<QString, QString> QgsPhongMaterialSettings::toExportParameters() const
+{
+  QMap<QString, QString> parameters;
+  parameters[ QString( "Kd" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mDiffuse.redF() ).arg( mDiffuse.greenF() ).arg( mDiffuse.blueF() );
+  parameters[ QString( "Ka" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mAmbient.redF() ).arg( mAmbient.greenF() ).arg( mAmbient.blueF() );
+  parameters[ QString( "Ks" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mSpecular.redF() ).arg( mSpecular.greenF() ).arg( mSpecular.blueF() );
+  parameters[ QString( "Ns" ) ] = QStringLiteral( "%1" ).arg( mShininess );
```
```suggestion
  parameters[ QStringLiteral( "Kd" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mDiffuse.redF() ).arg( mDiffuse.greenF() ).arg( mDiffuse.blueF() );
  parameters[ QStringLiteral( "Ka" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mAmbient.redF() ).arg( mAmbient.greenF() ).arg( mAmbient.blueF() );
  parameters[ QStringLiteral( "Ks" ) ] = QStringLiteral( "%1 %2 %3" ).arg( mSpecular.redF() ).arg( mSpecular.greenF() ).arg( mSpecular.blueF() );
  parameters[ QStringLiteral( "Ns" ) ] = QStringLiteral( "%1" ).arg( mShininess );
```

nyalldawson reviewed (commented):
Some minor code convention notes and qt related coding tips, otherwise looks good!

NEDJIMAbelgacem committed:
Move initialization to header, erase parent_name from export object

NEDJIMAbelgacem committed:
Code cleanup

NEDJIMAbelgacem committed:
erase inheritance from QObject

NEDJIMAbelgacem committed:
change to QgsDebugMsg

NEDJIMAbelgacem committed:
Moved finding geometries from parseVectorLayerEntity to each symbol class

NEDJIMAbelgacem committed:
Add SIP_SKIP to QgsPolygon3DSymbol::exportGeometries and its overloads

NEDJIMAbelgacem committed:
Merge master and fix merge conflict

NEDJIMAbelgacem committed:
Add toExportParameters() to DummyMaterialSettings test

wonder-sk reviewed (approved):
I am happy with the current implementation - great work @NEDJIMAbelgacem  :tada:

nirvn commented:
Wouhou!
